server/server.go

```
@@ -42,72 +55,362 @@ func HandleWebsocketConnection(serverHTTPResponse http.ResponseWriter, clientHTT
-	// Create a byte sequence that we would ask the client to sign
-	byteSequenceToSign := []byte("hi")
-	// Initialize the Auth protobuf structure
-	protobufAuth := backendProtobuf.Auth{}
-	// Fill in the byte sequence that we would ask the client to sign
-	protobufAuth.ToSign = byteSequenceToSign
+func processMessage(websocketConnection *websocket.Conn, authenticatedIdentityPublicKeyHexClient string) error {
+	// Initialize an empty variable to hold the protobuf message
+	var messageFromClientProtobuf backendProtobuf.BackendMessage
+	// Read a message from a client over the websocket connection
+	_, messageFromClientBytes, readMessageErr := websocketConnection.ReadMessage()
+	if readMessageErr != nil {
+		// If there is an error while reading the response from the client, return the error
+		return readMessageErr
+	}
+	// Unmarshal the protobuf bytes from the message we received into our protobuf message structure
+	protoUnmarshalErr := proto.Unmarshal(messageFromClientBytes, &messageFromClientProtobuf)
+	if protoUnmarshalErr != nil {
+		// If there is an error while unmarshaling the response from the client, return the error
+		return protoUnmarshalErr
+	}
+	// Outer master switch, cases are valid if they are true
+	switch {
+	// In case a message has both a request and a response
+	case messageFromClientProtobuf.Request != nil && messageFromClientProtobuf.Response != nil:
+		// Return an error as it's not allowed for a message to have both a response and a request at the same time
+		return errors.New("A message can’t have a response and a request at the same time")
+	// If the message is a request
+	case messageFromClientProtobuf.Request != nil:
+		// Inner switch in case we have a Request from client, cases are valid if they are true
+		switch {
+		case messageFromClientProtobuf.Request.Messages != nil:
+			handleMessageFromClient(websocketConnection, messageFromClientProtobuf.Request.Messages)
```

the returned error is ignored

---

Noted.

All errors that don't break the websocket connection, will be sent back to client.

For the errors that break the websocket connection, they will be logged.